### ISSUE TITLE
WAI reverse proxy eats + in query string

### ISSUE THREAD

**Issue by tdammers** [fpco/http-reverse-proxy#33]
I have an application that passes inputs to itself via query string parameters. These inputs often contain the "+" character (they are brainfuck programs, so this is absolutely inevitable).

When I run the application standalone (or talk to it directly via the local port it listens on, e.g. `http://localhost:5500/`), those "+" characters get percent-encoded into %2B by the browser's handling of GET forms (which is correct and desirable), and the server correctly parses the "+" back out and feeds it back into the form. E.g., writing the program `+++++++++++++++++++++++++++++++++.` into the form and submitting it produces the following URL: `http://localhost:5500/?program=%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B.&input=`. The program is executed and reflected correctly (it prints "!", and the original program appears in the form as it was entered). Debug-tracing the `program` request parameter (read using `param` from `scotty`) shows that it is parsed as `"+++++++++++++++++++++++++++++++++."`, so we can see that everything works correctly.

However, running the application behind a WAI reverse proxy that maps incoming requests by subdomain, using a stateful lookup table maintained in an `MVar` by the parent application, I get different behavior. The URL still looks correct, e.g.: `http://pxzfeaiiuzbmfvbb.redacted.example.org/?program=%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B%2B.&input=`. Same query string. However, the reflected program appears as `                                 .`, and prints as `"                                 ."` in the debug log, suggesting that the query string doesn't survive the reverse proxying unchanged. I fully expect the reverse proxy to potentially change the encoding of the string - e.g., it would be valid to percent-encode spaces previously given as literal `+` in the URL. But what seems to be happening here is that the reverse proxy parses the request, converting `%2B` into `+` (which is correct), but then fails to reverse the decoding when passing it to the proxied child application, keeping the `+` as such instead of percent-encoding it. The child application then sees a literal `+` in the encoded query string parameter, and decodes it into a space (which is correct behavior in itself).

In short:

- Setup: use `waiProxyTo` to reverse-proxy requests by mapping subdomains to ports (`WPRProxyDest`). Payload application reads brainfuck programs from query string and reflects those in HTML.
- Expected: `+` characters in cleartext input are encoded as `%2B` in the client-side URL, presented to the proxied payload app as `%2B`, and decode and reflect as `+`.
- Observed: `+` characters in cleartext input are encoded as `%2B` in the client-side URL, but reflect as ` ` (space).

**Comment by snoyberg** [comment]
Best bet here on a resolution would be a PR that you've tested and which resolves the issue in your environment. In order to work on in here, best next step would be a reproducing test case demonstrating the issue.

**Comment by tdammers** [comment]
Never mind, I was being too quick on the judgment. Turns out http-reverse-proxy isn't the only variable that differs between the "works" and "doesn't work" scenarios; the actual culprit, I found, is the lighttpd `proxy` module running in front of the reverse-proxying host.

Configuring a lighttpd reverse proxy to talk to the reverse-proxied child directly reproduces the problem, and doesn't involve `http-reverse-proxy` at all.

Sorry to have wasted your time here.

**Comment by snoyberg** [comment]
No worries!
